File: Heaps/LargeEleInStream/stream.cpp
#include <queue>
using namespace std;

class KthLargest {
public:
    priority_queue<int,vector<int>,greater<int>>pq;
    int k;
    KthLargest(int k, vector<int>& nums) {
        this->k = k;
        int n = nums.size();
        for(int i=0;i<n;i++){
            pq.push(nums[i]);
            if(pq.size()>k) pq.pop();
        }
    }
    
    int add(int val) {
        pq.push(val);
        if(pq.size()>k) pq.pop();
        if(pq.size()<k)return -1;
        return pq.top();
    }
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

class Solution {
  public:
    vector<int> kthLargest(vector<int>& arr, int k) {
        // code here
        vector<int>nums;
        KthLargest kLarge(k,nums);
        vector<int>ans;
        int n = arr.size();
        for(int i=0;i<n;i++){
            ans.push_back(kLarge.add(arr[i]));
        }
        return ans;
    }
};

File: Heaps/LargeEleInStream/stream.py
from typing import List
import heapq

class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.pq = []
        self.k = k
        n = len(nums)
        for i in range(n):
            heapq.heappush(self.pq,nums[i])
            if len(self.pq) > self.k:
                heapq.heappop(self.pq)  

    def add(self, val: int) -> int:
        heapq.heappush(self.pq,val)
        if len(self.pq) > self.k:
            heapq.heappop(self.pq)
        if len(self.pq)<self.k: return -1
        return self.pq[0]


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)


class Solution:
    def kthLargest(self, arr, k):
        # code here 
        
        kLarge = KthLargest(k,[])
        n = len(arr)
        ans = []
        
        for i in range(n):
            ans.append(kLarge.add(arr[i]))
        
        return ans